### ISSUE TITLE
[BUG] speculative decode test sometimes fails (SpeculativeDecodingTests)

### ISSUE THREAD

**Issue by davidkoski** [ml-explore/mlx-swift-lm#315]
I observed that the SpeculativeDecodingTests sometimes fails.

To investigate I tried making the test more deterministic:

- I decreased the model size
- I used fixed random state so that the main and draft model were _identical_

```swift
        let modelConfig = Gemma3TextConfiguration(
            modelType: "text",
            hiddenSize: 8, hiddenLayers: 8, intermediateSize: 64,
            attentionHeads: 4, headDim: 8,
            rmsNormEps: 0.00001, vocabularySize: 100, kvHeads: 4,
            ropeTheta: 1_000_000, ropeLocalBaseFreq: 10_000,
            ropeTraditional: false, queryPreAttnScalar: 256,
            slidingWindow: 512, slidingWindowPattern: 6,
            maxPositionEmbeddings: 2048
        )

        let mainModel = withRandomState(.init(seed: 0)) {
            Gemma3TextModel(modelConfig)
        }
        let mainContext = ModelContext(
            configuration: processor.configuration,
            model: mainModel,
            processor: processor,
            tokenizer: processor.tokenizer
        )

        let draftModel = withRandomState(.init(seed: 0)) {
            Gemma3TextModel(modelConfig)
        }
        let draftContext = ModelContext(
            configuration: processor.configuration,
            model: draftModel,
            processor: processor,
            tokenizer: processor.tokenizer
        )
```

I verified that the weights were the same:

```swift
        print("mm: \(mainModel.parameters().mapValues { sum($0).item(Float.self) }.reduce(0, +))")
        print("dm: \(draftModel.parameters().mapValues { sum($0).item(Float.self) }.reduce(0, +))")
```

The failures reproduce at all sizes and with or without logit processors, but it seems to happen more frequently with the larger draft token sizes (but again, still fails at 2):

```swift
    @Test(arguments: [48], [false])
    func `Speculative decoding matches default token generation`(
        numDraftTokens: Int,
        withLogitProcessor: Bool
    ) async throws {
```

I put the inside of the test in a loop -- does it fail with the _same_ model parameters?  It does:

```
iter 0
iter 1
iter 2
iter 3
iter 4
􀢄 Test "Speculative decoding matches default token generation" recorded an issue with 2 arguments numDraftTokens → 48, withLogitProcessor → false at SpeculativeDecodingTests.swift:94:13: Expectation failed: (normalTokens → [93, 84, 97, 14, 84, 97, 71, 84, 53, 84, 98, 71, 84, 84, 84, 84, 68, 40, 68, 40, 90, 84, 84, 57, 57, 57, 56, 22, 28, 90, 68, 84]) == (speculativeTokens → [93, 84, 97, 14, 84, 97, 71, 84, 53, 84, 98, 71, 84, 84, 84, 84, 68, 40, 68, 40, 90, 84, 84, 57, 57, 57, 56, 22, 28, 90, 68, 68])
􀅺 inserted [84], removed [68]
iter 5
iter 6
iter 7
iter 8
iter 9
􀢄 Test "Speculative decoding matches default token generation" recorded an issue with 2 arguments numDraftTokens → 48, withLogitProcessor → false at SpeculativeDecodingTests.swift:94:13: Expectation failed: (normalTokens → [44, 44, 3, 45, 3, 45, 86, 49, 68, 22, 53, 22, 22, 45, 49, 90, 45, 22, 39, 44, 28, 45, 17, 22, 22, 22, 45, 53, 68, 62, 17, 44]) == (speculativeTokens → [44, 44, 3, 45, 45, 3, 45, 86, 49, 90, 45, 90, 45, 68, 22, 60, 49, 90, 71, 45, 27, 62, 27, 62, 27, 68, 52, 17, 53, 17, 90, 68])
􀅺 inserted [53, 22, 22, 45, 22, 39, 44, 28, 45, 22, 22, 22, 45, 62, 17, 44], removed [45, 90, 45, 90, 45, 60, 71, 27, 62, 27, 62, 27, 68, 52, 17, 90]
iter 10
iter 11
􀢄 Test "Speculative decoding matches default token generation" recorded an issue with 2 arguments numDraftTokens → 48, withLogitProcessor → false at SpeculativeDecodingTests.swift:94:13: Expectation failed: (normalTokens → [45, 90, 22, 3, 3, 22, 38, 86, 28, 90, 48, 90, 45, 90, 45, 90, 45, 90, 45, 90, 22, 60, 21, 28, 28, 28, 28, 28, 28, 28, 28, 28]) == (speculativeTokens → [45, 90, 22, 22, 45, 28, 28, 90, 22, 71, 28, 28, 28, 54, 17, 49, 28, 28, 53, 27, 17, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28, 28])
􀅺 inserted [3, 3, 38, 86, 90, 48, 90, 45, 90, 45, 90, 45, 90, 45, 60, 21], removed [45, 28, 71, 28, 28, 28, 54, 17, 49, 28, 28, 53, 27, 17, 28, 28]
iter 12
iter 13
```

The `normalTokens` are not identical because the TestTokenizer produces random tokens.

If I keep the `LMInput` fixed then it never fails on the same input and always fails if we have a "bad" input.  For example with these tokens:

```
[92, 85, 2, 95, 55, 7, 94, 42]
```

it always fails (with seed 0):

```
iter 0
[22, 71, 90, 45, 28, 38, 45, 90, 45, 90, 45, 28, 38, 45, 90, 45, 28, 38, 45, 90, 45, 90, 45, 90, 71, 90, 71, 27, 38, 17, 60, 28]
􀢄 Test "Speculative decoding matches default token generation" recorded an issue with 2 arguments numDraftTokens → 48, withLogitProcessor → false at SpeculativeDecodingTests.swift:96:13: Expectation failed: (normalTokens → [22, 71, 90, 45, 28, 38, 45, 90, 45, 90, 45, 28, 38, 45, 90, 45, 28, 38, 45, 90, 45, 90, 45, 90, 71, 90, 71, 27, 38, 17, 60, 28]) == (speculativeTokens → [22, 71, 90, 45, 28, 38, 45, 90, 45, 90, 45, 28, 38, 45, 90, 45, 90, 71, 90, 71, 27, 71, 27, 22, 69, 38, 45, 49, 38, 45, 49, 38])
􀅺 inserted [28, 38, 45, 45, 90, 45, 90, 17, 60, 28], removed [71, 27, 22, 69, 45, 49, 38, 45, 49, 38]
```

**Comment by davidkoski** [comment]
@petrukha-ivan FYI in case you have any ideas

**Comment by davidkoski** [comment]
In the token iterator (draft size 48, numDraft is 32):

```swift
        for i in 0 ..< numDraft {
            guard mainTokensList[i] == draftTokensList[i] else {
                break
            }

            processor?.didSample(token: draftTokens[i])
            pendingTokens.append(mainTokensList[i])
            accepted += 1
        }
```

it diverges at 16, which is maybe surprising since the models are identical, though because of float16 math it is _possible_.  The exact order and tiling, e.g. for matrix multiply, can produce slight differences.

```swift
(lldb) p i
(Int) 16
(lldb) p mainTokensList
([Int]) 33 values {
  [0] = 22
  [1] = 71
  [2] = 90
  [3] = 45
  [4] = 28
  [5] = 38
  [6] = 45
  [7] = 90
  [8] = 45
  [9] = 90
  [10] = 45
  [11] = 28
  [12] = 38
  [13] = 45
  [14] = 90
  [15] = 45
  [16] = 90  <--- HERE
  [17] = 38
  [18] = 45
  [19] = 90
  [20] = 45
  [21] = 90
  [22] = 45
  [23] = 90
  ...
}
*** Some of the displayed variables have more members than the debugger will show by default. To show all of them, you can either use the --show-all-children option to dwim-print or raise the limit by changing the target.max-children-count setting.
(lldb) p draftTokensList
([Int]) 32 values {
  [0] = 22
  [1] = 71
  [2] = 90
  [3] = 45
  [4] = 28
  [5] = 38
  [6] = 45
  [7] = 90
  [8] = 45
  [9] = 90
  [10] = 45
  [11] = 28
  [12] = 38
  [13] = 45
  [14] = 90
  [15] = 45
  [16] = 28 <--- HERE
  [17] = 38
  [18] = 45
  [19] = 90
  [20] = 45
  [21] = 90
  [22] = 45
  [23] = 90
  ...
}
```

so the next step is:

```swift
        // Always emit the main model's token at position `accepted`
        // (either the correction token or the bonus token if all drafts matched)
        let finalToken = mainTokens[accepted ... accepted]
        processor?.didSample(token: finalToken)
        pendingTokens.append(mainTokensList[accepted])
```

we take the 90 from the main tokens but this is the point of divergence.

A few thoughts/questions:

- should we take the token from main when we diverge?  I think so -- in normal cases the main model is more robust
- the fact that there is numerical instability in float16 -- is my reasoning correct?  I think so, you can observe this when summing all the weights for example
- but the fact that once the repro situation is set up it is 100% consistent is curious -- how do I understand the float16 instability?  is my mental model incorrect?

Is there a way to make the test 100% reliable?

**Comment by petrukha-ivan** [comment]
Hey @davidkoski, I took a look at this and there are a few interesting things here.

> should we take the token from main when we diverge? I think so -- in normal cases the main model is more robust

Yes, that's the standard speculative decoding algorithm: on rejection we sample from the main distribution and discard the draft suffix. Doing anything else changes the final output and makes speculative decoding unreliable.

> the fact that there is numerical instability in float16 -- is my reasoning correct? I think so, you can observe this when summing all the weights for example

Indeed float16 has numerical instability, but I'm not sure it has a significant impact here. I traced the path and all weights and intermediate activations are in float32. Did you have float16 or bfloat16 arrays at some point?

> but the fact that once the repro situation is set up it is 100% consistent is curious -- how do I understand the float16 instability? is my mental model incorrect?

This one is curious because I don't have a repro with the setup you provided: I changed the model configuration, fixed the random state, provided a fixed `LMInput`, and the test still passes. I also tried running the test repeatedly with random models and inputs, and it didn't fail a single time across 300 runs. I also checked GitHub Actions history and didn't find any speculative decoding test failures -- did you observe such behaviour on CI?

What I noticed is that at divergence position 16 there are two high-probability logits with almost identical values, exactly at token positions 28 and 90. Can you execute this in `TokenIterator.convertToToken(logits:)` with a conditional breakpoint `tokenCount == 15` to see how we get the token at position 16:

```
(lldb) po logits[0, MLXArray([28, 90])]
array([1.80186, 1.8023], dtype=float32)
```

And this one in `SpeculativeTokenIterator.speculateRound()` just after we get `let verifyLogits = ...` to see the difference in raw logits between paths:

```
(lldb) po verifyLogits[16, MLXArray([28, 90])]
array([1.80187, 1.8023], dtype=float32)
```

What values do you have?

I have nearly identical values, but even the 0.00001 (1.80186 vs 1.80187) difference shows that something can drift here. Still, the token at position 90 is higher in both cases, so `ArgMaxSampler` chooses 90:

```
(lldb) po verifyLogits[16, MLXArray([28, 90])].argMax()
array(1, dtype=uint32)
```

The only way for me to select another one is to lower precision to bfloat16 specifically, because float16 still works the same way as float32 here. So with a bfloat16 downcast I get identical numbers, but as I said before, I don't observe any downcast to float16 or bfloat16 in the forward pass:

```
(lldb) po verifyLogits[16, MLXArray([28, 90])].asType(.bfloat16)
array([1.80469, 1.80469], dtype=bfloat16)
```

And when they're identical, the sampler chooses the first index:

```
(lldb) po verifyLogits[16, MLXArray([28, 90])].asType(.bfloat16).argMax()
array(0, dtype=uint32)
```

I don't think there's a problem or bug with speculative decoding or the tests. It looks like in your environment the autoregressive path `1 -> 2 -> 3 -> 4 -> 5` is not guaranteed to be equal to the batched path `[1, 2, 3, 4] -> 5`, even with the same model and configuration. That's why in the speculative decoding path, when you use an identical model as drafter and main, your `draftTokensList` has token 28 at position 16 -- the same as what you get with normal decoding, because both use autoregressive evaluation. And `mainTokensList` has token 90 at position 16 because it accepts the drafts and processes them in a batch. Interesting that on my machine both paths land on token 90, which is why I can't repro and dig deeper on my own. Let me know what those breakpoints show on your machine and I'll take another look.

**Comment by davidkoski** [comment]
> This one is curious because I don't have a repro with the setup you provided: I changed the model configuration, fixed the random state, provided a fixed LMInput, and the test still passes. I also tried running the test repeatedly with random models and inputs, and it didn't fail a single time across 300 runs. I also checked GitHub Actions history and didn't find any speculative decoding test failures -- did you observe such behaviour on CI?

Yes the first time I saw it fail was on CI, but I agree, I can't find it now.  Let me make a branch to see if I can trigger it.

I suggest:

- use a fixed random seed for the models
- run it in a loop with whatever parameters you want -- it triggered for all combinations for me
- use different LMInputs each time (e.g. do the "tokenization" in the loop)
- you can print the tokens and then when it fails that should be the repro case for you (perhaps it differs slightly by OS or hardware?)

Originally I use the Xcode feature to run the test in a loop and then tried things to make it fail faster / more reliably.  It seemed to fail around 5% of the time on my laptop.

**Comment by davidkoski** [comment]
> Indeed float16 has numerical instability, but I'm not sure it has a significant impact here. I traced the path and all weights and intermediate activations are in float32. Did you have float16 or bfloat16 arrays at some point?

Ah I assumed, but you are correct -- all float 32!

> What I noticed is that at divergence position 16 there are two high-probability logits with almost identical values, exactly at token positions 28 and 90. Can you execute this in TokenIterator.convertToToken(logits:) with a conditional breakpoint tokenCount == 15 to see how we get the token at position 16:

```
(lldb) po logits[0, MLXArray([28, 90])]
array([1.80468, 1.80448], dtype=float32)

vs your result:

array([1.80186, 1.8023], dtype=float32)
```

> And this one in SpeculativeTokenIterator.speculateRound() just after we get let verifyLogits = ... to see the difference in raw logits between paths:

```
(lldb) po verifyLogits[16, MLXArray([28, 90])]
array([1.80324, 1.80435], dtype=float32)

vs your result:

array([1.80187, 1.8023], dtype=float32)
```

Ah, there it is -- the order is swapped for me.

OK, so I have an M5 on my desk and if I _disable_ NAX then I get the same result as you.  That makes some sense -- these batched evals will use the NAX path.

If I keep the NAX disabled then at least my repro case fails to repro.  I will continue to investigate, but those are good clues!

**Comment by davidkoski** [comment]
OK, I think I tricked myself.  I think I saw this failure:

- https://github.com/ml-explore/mlx-swift-lm/actions/runs/26472502733/job/77951038593

which is `testQwen2ConversionImage` (and was somewhat expected based on the test).  The last line of the log is some longish array being printed.  When I ran unit tests on this branch I bet that is when I saw this failure and I just figured it must have also failed in CI, but it did not.

**Comment by davidkoski** [comment]
There are two possible approaches here:

- use `MLX_ENABLE_TF32=0` to disable using the NAX on float32
- maybe (I haven't tried this yet) we can just use float16 parameters -- that would be closer to what actual models use.  the problem isn't the precision as much as the _difference_ in precision in NAX vs non-NAX

But I think the mystery is solved -- thank you for your help @petrukha-ivan !

**Comment by petrukha-ivan** [comment]
@davidkoski Nice find! The NAX vs non-NAX difference explains why I couldn't repro on my M3. Both possible approaches sound good. NAX uses lower precision internally, so either disabling it in tests or moving the model to float16 should fix the issue. Glad I could help -- happy to take another look if anything else comes up!
